commit 48499f2987bd32b970a043e8ac56c82a300a7c67 (refs/changes/95/2295/3)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-07-27T14:37:31-07:00 (2 years, 2 months ago)
    
    IMGAPI-636 docker pull retry not working for v2.1 images
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/CHANGES.md b/CHANGES.md
index bf0cae5..8860f44 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # IMGAPI changelog
 
+## 4.0.2
+
+- IMGAPI-636: docker pull retry not working for v2.1 images
+
 ## 4.0.1
 
 - IMGAPI-635: imgapi crashing on 'docker pull holzi/quine-relay:latest'
diff --git a/lib/docker.js b/lib/docker.js
index 392e04f..7e43df3 100644
--- a/lib/docker.js
+++ b/lib/docker.js
@@ -465,7 +465,7 @@ function dockerDownloadAndImportLayer(opts, callback) {
 
         stream.on('data', function (chunk) {
             if (ctx.downloadsCanceled) {
-                stream.destroy();
+                resp.destroy();
                 progressStatus('Aborted');
                 return;
             }
@@ -494,6 +494,16 @@ function dockerDownloadAndImportLayer(opts, callback) {
             finish(errors.wrapErrorFromDrc(streamErr));
         });
 
+        if (resp !== stream) {
+            // IMGAPI-636: Listen for errors on the original response object,
+            // as the stream is a buffer-peek-stream (for v2.1 image layers),
+            // and drc will emit the error on the original stream (from the
+            // createBlobReadStream call).
+            resp.on('error', function (streamErr) {
+                finish(errors.wrapErrorFromDrc(streamErr));
+            });
+        }
+
         stor = app.chooseStor(newImage);
         stor.storeFileFromStream({
             image: newImage,
diff --git a/package.json b/package.json
index e5c722a..2832144 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgapi",
   "description": "Image API to manage images for SDC 7",
-  "version": "4.0.1",
+  "version": "4.0.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
